From bc2caff9cdef8a16297a2cb196306f88fa5ea2c2 Mon Sep 17 00:00:00 2001
From: Andy Green <andy@openmoko.com>
Date: Thu, 2 Oct 2008 12:40:10 +0100
Subject: [PATCH] revert-fix-build-with-no-config_mmc-glamo-resume-callback-part.patch

Thomas White noticed that the recent patch from Andrzej cleaning up a
nasty cast in the resume_dependency stuff for Glamo broke resume.  The
problem was that the wrong resume callback was arrived at by the new
code, the one in the device's device_driver struct rather than the
struct platform_driver that actually holds the right pointer.

Since this code will be gone in 2.6.26, I reverted this part of Andrzej's
patch, tidying the cast a bit anyway.

Reported-by: Thomas White <taw27@cam.ac.uk>
Signed-off-by: Andy Green <andy@openmoko.com>
---
 arch/arm/mach-s3c2440/mach-gta02.c |    4 +++-
 drivers/mfd/glamo/glamo-mci.c      |    3 ++-
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/arch/arm/mach-s3c2440/mach-gta02.c b/arch/arm/mach-s3c2440/mach-gta02.c
index 435378a..467c417 100644
--- a/arch/arm/mach-s3c2440/mach-gta02.c
+++ b/arch/arm/mach-s3c2440/mach-gta02.c
@@ -1417,8 +1417,10 @@ static int gta02_glamo_mci_all_dependencies_resumed(struct platform_device *dev)
 
 static void gta02_glamo_mci_suspending(struct platform_device *dev)
 {
+	int glamo_mci_resume(struct platform_device *dev);
+	
 #if defined(CONFIG_MFD_GLAMO_MCI) && defined(CONFIG_PM)
-	resume_dep_glamo_mci_pcf.callback = (void *)dev->dev.driver->resume;
+	resume_dep_glamo_mci_pcf.callback = (void (*)(void *))glamo_mci_resume;
 	resume_dep_glamo_mci_pcf.context = (void *)dev;
 	pcf50633_register_resume_dependency(pcf50633_global,
 						     &resume_dep_glamo_mci_pcf);
diff --git a/drivers/mfd/glamo/glamo-mci.c b/drivers/mfd/glamo/glamo-mci.c
index d5099a0..fb6224d 100644
--- a/drivers/mfd/glamo/glamo-mci.c
+++ b/drivers/mfd/glamo/glamo-mci.c
@@ -1006,7 +1006,7 @@ static int glamo_mci_suspend(struct platform_device *dev, pm_message_t state)
 	return ret;
 }
 
-static int glamo_mci_resume(struct platform_device *dev)
+int glamo_mci_resume(struct platform_device *dev)
 {
 	struct mmc_host *mmc = platform_get_drvdata(dev);
 	struct glamo_mci_host 	*host = mmc_priv(mmc);
@@ -1025,6 +1025,7 @@ static int glamo_mci_resume(struct platform_device *dev)
 
 	return ret;
 }
+EXPORT_SYMBOL_GPL(glamo_mci_resume);
 
 #else /* CONFIG_PM */
 #define glamo_mci_suspend NULL
-- 
1.5.6.5

